src/client/systems/ClientGameController.cpp

```diff
@@ -132,73 +143,106 @@ void ClientGameController::postUpdate(sf::Int32 update_time) {
+}
 
-    // Prevent the client from drifting too far from server
-    tick_delta = std::min(tick_delta, GAME_SETTINGS.REPLAY_TICK_DELTA_THRESHOLD);
-
-    // Loop through ticks that need to be replayed and apply client input from cache if present
-    for (int i = 0; i < tick_delta; ++i) {
-        uint32_t replay_tick = server_tick + i;
-        if (m_tickToInput.count(replay_tick) > 0) {
-            InputDef::ClientInputAndTick client_input_and_tick = m_tickToInput[replay_tick];
-            auto pi = InputDef::PlayerInputs(m_inputController->getPlayerInputs(
-                m_playerId, client_input_and_tick.ri, client_input_and_tick.ui));
-            GameController::computeGameState(pi, GameController::MIN_TIME_STEP);
-        } else {
-            // If we don't have input for this tick pass in empty inputs
-            auto pi = InputDef::PlayerInputs();
-            GameController::computeGameState(pi, GameController::MIN_TIME_STEP);
+/**
+ * If client is ahead of the smallest server game state in the buffer, stay at the same tick but
+ * interpolate to the next available game state in the buffer. If there are none available just keep
+ * the game state the same. Note that the tick stays the same in this case so that we fall back
+ * behind to being behind the smallest server game state in the buffer.
+ */
+void ClientGameController::ahead(uint32_t client_tick, uint32_t smallest_server_tick,
+                                 std::map<uint32_t, GameState>& game_state_buffer) {
+    std::map<uint32_t, GameState>::iterator game_state_buffer_it = game_state_buffer.begin();
+    ++game_state_buffer_it;
+    while (game_state_buffer_it != game_state_buffer.end()) {
+        uint32_t next_server_tick = game_state_buffer_it->first;
+        if (client_tick < next_server_tick) {
+            const GameState& next_server_game_state = game_state_buffer_it->second;
+            uint32_t next_tick = client_tick;
+            // Keep the same tick but interpolate to the next available server game state.
+            interpolateGameState(client_tick - 1, next_tick, next_server_tick,
+                                 next_server_game_state);
+            m_stallGameStateMetric.pushCount(client_tick - smallest_server_tick);
+            return;
         }
+        ++game_state_buffer_it;
     }
 
-    // Remove cached ticks older than the applied server game state's tick. We won't need them for
-    // future replays because we never apply stale game states.
-    for (auto it = m_tickToInput.begin(); it != m_tickToInput.end();) {
-        if (it->first < server_tick) {
-            m_tickToInput.erase(it++);
-        } else {
-            ++it;
-        }
+    // If we reach here then we don't have any game states to interpolate to. Keep the same tick and
+    // game state.
+    m_stallGameStateMetric.pushCount(client_tick - smallest_server_tick);
+    return;
```

Comment:
Yes
Even when we stall, we use the next game state we have to interpolate the same tick we're currently on (in case our prediction from before that got us in the state we are was wrong). 
We only want to do this for the next game state. If we don't return then we'll keep doing this for every game state head of us in the buffer.

now that i think of it tho, maybe we should using the _latest_ game state we have, that's the most accurate.....

Reply:
I guess I mean this is the last statement in the function... i thought functions returned null by default if you didn't explicitly put a statement, so im wondering if this return statement is unnecessary